README.md
# OpenSpending - WordPress plugin

WordPress plugin to make openspendingjs visualisations easy to add to WordPress posts and pages.

## Installation

1. Add the openspending directory into your WordPress instance's plugin
   directory (usually *wp-content/plugins/*).
2. Go to your admin panel, navigate to plugins, find OpenSpending and
   activate it.

## How to Use

The OpenSpending WordPress plugin operates using the standard WordPress *shortcodes*, namely:

    [openspending]

Three attributes **must** be defined:

* *type* - The type of visualisation to use
* *dataset* - The dataset on openspending.org to use
* *drilldowns* - Dimensions of the dataset to drill down into

If you would like to add a *treemap* for the dimensions *from* and *to* in the [*ukgov-25k-spending* dataset](http://openspending.org/ukgov-25k-spending) on OpenSpending the only thing you need to add to your post or page is:

    [openspending type="treemap" dataset="ukgov-25k-spending" dimensions="from,to"]

## Supported Visualisations

* Treemap (type: "treemap")
* Bubbletree (type: "bubbletree")
    * Supported icons/colors for COFOG

## FAQ

###How to change the color of the main bubbles in the bubble chart?
This would require javascript overwrite. It would be difficult to add onto wordpress ui.

###How to change line height of the bubble titles?
Would require css overwrite. It would difficult to add onto wordpress ui.

###How to display amounts in Millions instead of Billions?
Would require javascript overwrite (is possible, but difficult to add onto wordpress ui)

###How to change the assignment of bubble background images?
Wordpress often stops users from adding javascripts files so it will require code changes to wordpress itself (or the theme)

###How to add a secondary currency, like US dollars and have it be shown in the bubbles or in the pop-up window?
Unfortunately you cannot do this at the moment in OpenSpending. This would need a customised improvement.You can however do a separate budget in USD dollars and visualise that.


CHANGELOG
# v0.4.2

* Users with non-standard wordpress setup now won't have to muck about with any files

# v0.4.1

* Bug fix: Bubbletree can now work with non-string identified data

# v0.4.0

* Users can now create bubbletrees
* Supported icons/colors for COFOG levels

# v0.3.0

* Users can now select a year to display in the visualisation
* Dimensions (and years) are sorted alphabetically for easier use
* Fine tuning of look and feel

# v0.2.1

* Bug fix relating to drilldowns
* Documentation for the user interface added

# v0.2

* OpenSpending button added to the WordPress editor
* User interface to create the shortcode with attributes

# v0.1

* Initial release!
* Possibility to add shortcode [openspending] to posts/pages
* Shortcode attributes must be type, dataset, and drilldowns


openspending/openspending.php
<?php
/*
Plugin Name: OpenSpending
Plugin URI: http://tryggvib.github.io/openspending-wordpress-plugin/
Description: Easily add <a href="http://openspending.org" title="OpenSpending - Mapping the Money">OpenSpending</a> visualisations to your blog or pages using the openspending shortcode. This means that you can easily map the money by loading a dataset into <a href="http://openspending.org" title="OpenSpending - Mapping the Money">OpenSpending</a> (if it isn't already there) and dropping a shortcode into your WordPress!
Author: Open Knowledge Foundation
Version: 0.4.2
Author URI: http://okfn.org/
License: GPLv2 or later
*/

/*  Copyright 2013 Open Knowledge Foundation ( http://okfn.org )

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 2, as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

/* OpenSpending plugin class
   Adds a shortcode handler for [openspending] that needs at least three
   attributes:
   * 'type': The type of visualisation
             e.g. treemap, bubbletree
   * 'dataset': What dataset on OpenSpending it should visualize,
                e.g. ukgov-finances-cra
   * 'drilldowns': Dimensions of the data to drilldown into,
                   e.g. from,to
*/
class OpenSpending {

    // Variable set when shortcode is found to print stylesheets and scripts
    static $add_style_and_script;

    // Set hooks for this plugin
    static function init() {
        // Add popup asset hack to the init hook
        add_action( 'init', array( __CLASS__, 'openspending_popup_assets' ) );

        // This plugin searches for the openspending shortcode which creates
        // as visualisation
        add_shortcode('openspending', array(__CLASS__, 'add_visualisation'));

        // Add triggers to register scripts and print them
        add_action('init', array(__CLASS__, 'register_style_and_script'));
        add_action('wp_footer', array(__CLASS__, 'print_style_and_script'));

        add_action('init', array(__CLASS__, 'add_mce_button'));
    }

    // Function called when the shortcode is found in a post/page
    static function add_visualisation($atts) {
        // Extract the attributes for the shortcode
        extract( shortcode_atts( array(
            'type' => null,
            'dataset' => null,
            'drilldowns' => null,
            'year' => null,
	), $atts ) );

        // Type, Dataset and Drilldowns must be defined
        if ( $type == null || $dataset == null || $drilldowns == null )
            return "<emph>Badly formed OpenSpending visualisation</emph>";

        // Set variable that adds style and script to the page/post
        self::$add_style_and_script = true;

        $div_open = "<div class=\"{$type}\"" .
                    " data-dataset=\"{$dataset}\"" .
                    " data-drilldowns=\"{$drilldowns}\"";
        
        if ( $year != null)
            $div_open .= " data-year=\"{$year}\"";

        if ( $type == 'bubbletree' )
            $div_open .= " data-icons-path=\"".
                         plugins_url('openspending/svg/') .
                         "\"";

        return $div_open . "></div>";
    }

    // Register the stylesheet and javascript we might need to use
    static function register_style_and_script() {
        // Register the CSS stylesheet for inclusion in the page if needed
        wp_register_style('openspending', 
                          plugins_url('css/openspending.min.css', __FILE__));

        // Register the JS script for inclusion if needed.
        wp_register_script('openspending',
                           plugins_url('js/openspending.min.js', __FILE__),
                           array('jquery'),
                           null, true);
    }

    static function print_style_and_script() {
        // If the we don't have to add the style and script we just return
        if ( ! self::$add_style_and_script )
            return;

        // Add the stylesheets and the scripts. This way we add the stylesheet
        // into the body, not into the header, but we can safely assume that
        // it takes longer to make a request for the data than it is to parse
        // and apply the stylesheet
        wp_print_styles('openspending');
        wp_print_scripts('openspending');
    }

    static function add_mce_button() {
        if ( ! current_user_can('edit_posts') &&
             ! current_user_can('edit_pages') ) {
            return;
        }

        if ( get_user_option('rich_editing') == 'true' ) {
            add_filter( 'mce_external_plugins', array(__CLASS__,'add_plugin'));
            add_filter( 'mce_buttons', array(__CLASS__,'register_button'));
        }
    }


    static function register_button( $buttons ) {
        array_push( $buttons, "", "openspending" );
        return $buttons;
    }

    static function add_plugin( $plugin_array ) {
        $plugin_array['openspending'] = plugins_url('openspending/js/openspending-mce.js');
        return $plugin_array;
    }

    // Function called on init to check if openspending asset should be served
    // instead of the wordpress page itself. This is a "curvy" solution to the
    // problem raised by the plugin where a the tinymce popup page is loaded
    // with an iframe and wordpress functions aren't accessible. Instead of
    // loading wordpress core files (which are not necessarily stored in the
    // same location across wordpress instances) we make the assets available
    // this way.
    // This hack is based on a similar solution made by Marius Jensen (Clorith)
    static function openspending_popup_assets() {
        // We just check if the openspending-plugin-asset query variable is
        // present (on any page served by wordpress)
        if ( isset( $_GET['openspending-plugin-asset'] ) )
        {
            // If it is set the function must be one of these allowed functions
            $allowed_funcs = array('spinner', 'tinymce', 'styles', 'scripts');
            $func = $_GET['openspending-plugin-asset'];

            // Check if it's allowed and if it is we serve plain text content
            // where we call the class function (which prints out the asset)
            // and then we die() to avoid loading the wordpress page
            if ( in_array($func, $allowed_funcs) ) {
                header( 'Content-Type: text/plain' );
                call_user_func(array(__CLASS__, 'openspending_popup_' . $func));
                die();
            }
        }
    }

    // The following methods (openspending_popup_*) print out assets
    // available to the openspending tinymce popup

    static function openspending_popup_spinner() {
        // Spinner is the included wordpress spinner
        echo includes_url('images/wpspin.gif');
    }
    static function openspending_popup_tinymce() {
        // TinyMCE popup is the included one
        echo includes_url('js/tinymce/tiny_mce_popup.js');
    }
    static function openspending_popup_styles() {
        // We enqueue a few styles via the wordpress mechanism and tailor it
        // to the user then we print it out, again with the help of wordpress

        // Get the user's layout (for the colors)
        global $user_ID;
        $layout = get_user_meta($user_ID, 'admin_color', true);

        // Enqueue the stylesheets and scripts we'll be needing
        wp_enqueue_style( "colors-{$layout}" );
        wp_enqueue_style( "buttons" );
        wp_enqueue_style( 'ie' );
        wp_enqueue_style( 'popup-specific',
                          plugins_url('openspending/css/popup.css') );

        do_action('admin_print_styles');
    }
    static function openspending_popup_scripts() {
        // Load and print scripts (jquery) with the wordpress mechanism
        wp_enqueue_script( 'jquery' );
        do_action('admin_print_scripts');
    }
}

// Initialize our plugin class
OpenSpending::init();

?>


openspending/code/load-wordpress.php
<?php
/*  Copyright 2013 Open Knowledge Foundation ( http://okfn.org )

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 2, as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

/*
   Get the absolute path to wp-load.php so that we can load the WordPress
   functions and access them in our popup.
*/

// If you (the user) have put wp-content in a non-standard folder you need
// to put in the path and uncomment the following line
// define('WP_LOAD_PATH', '/path/to/wp-load.php');

// If WP_LOAD_PATH hasn't been define before we need to find it
if ( !defined('WP_LOAD_PATH') ) {
  // This file is in wordpressfolder/wp-content/plugins/openspending/js/
  // wp-load.php is in "wordpressfolder"
  // Get the directory name (and append a slash)
  $dir = dirname(dirname(dirname(dirname(dirname(__FILE__))))) . '/' ;

  // Check if wp-load.php is a file in the directory
  // If so we can define WP_LOAD_PATH
  if (file_exists( $dir . 'wp-load.php') )
    define( 'WP_LOAD_PATH', $dir . 'wp-load.php');
  // If not, we exit
  else
    exit("An error came up when loading WordPress");
}

// Load WordPress
require_once( WP_LOAD_PATH );
?>


openspending/code/add-shortcode.php
<?php
  /*  Copyright 2013 Open Knowledge Foundation ( http://okfn.org )

      This program is free software; you can redistribute it and/or modify
      it under the terms of the GNU General Public License, version 2, as 
      published by the Free Software Foundation.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program; if not, write to the Free Software
      Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
  */

  // This function takes care of loading the assets from WordPress via a
  // request to the site with a special query variable. The plugin itself
  // checks for this query variable and makes WordPress return the asset
  // if the plugin has defined it. To print the value it's enough to just
  // call print_asset('spinner'); to get the spinner asset (url to spinner)
  function print_asset($asset)
  {
    // The request submits the query var to the same server so we construct
    // the server protocol and name from the $_SERVER variable (we use
    // HTTP_HOST to include port number if there is one
    $home = ($_SERVER['HTTPS'] ? 'https://' : 'http://') . $_SERVER['HTTP_HOST'];
    // We print out the contents of the home page with the special plugin
    // query variable (with value we pass in)
    echo file_get_contents($home . '/?openspending-plugin-asset=' . $asset);
  }
?>
<!DOCTYPE html>
<head>
  <title>OpenSpending - Add your visualisation</title>
  <?php print_asset('styles'); ?>
  <?php print_asset('scripts'); ?>
  <script type="text/javascript" src="<?php print_asset('tinymce'); ?>"></script>
</head>
<body class="wp-core-ui">
<form id="openspending-shortcode">
  <p>
    <span class="type">
        <img src="../img/treemap-icon.png"><br>
        <label for="type">Treemap <input type="radio" name="type" value="treemap" checked></label>
    </span>
    <span class="type">
        <img src="../img/bubbletree-icon.png"><br>
        <label for="type">Bubbletree <input type="radio" name="type" value="bubbletree"></label>
    </span>
  </p>
  <p>
    <label for="dataset">Dataset:</label>
    <span class="url">http://openspending.org/</span>
    <input id="dataset" name="dataset" type="text" value="" placeholder="ukgov-25k-spending" />
  </p>
  <p id="drilldowns">
    <label for="drilldowns">Drilldowns</label><br/>
    <a class="button button-disabled" href="#" id="fetch-drilldowns">Fetch possible drilldowns</a>
    <div id="dimensions"></div>

    <label for="dimension" id="dimension-label" style="display:none;">Chosen drilldowns:</label><br/>
    <div id="chosen-dimensions"></div>
  </p>

  <p>
    <label for="year">Year <small>(optional)</small></label><br/>
    <a class="button button-disabled" href="#" id="fetch-years">Fetch available years</a>
    <div id="years"></div>
    <input id="year" type="hidden" name="year" value=""/>
  </p>
  <hr/>
  <p>
    <a id="add-visualisation" class="button button-primary button-large button-primary-disabled" href="#">
      Add visualisation
    </a>
  </p>
</form>

<script type="text/javascript">
  (function($) {
    // User presses "the big blue button" to add the shortcode to the page/post
    $('#add-visualisation').click(function (e) {
      // Get type of visualisation and do nothing if none is defined
      var visualisation = $('input[name=type]:checked').val();
      if (visualisation === '')
        return false;

      // Get dataset and do nothing if not defined
      var dataset = $('#dataset').val();
      if (dataset === '')
        return false;

      // Get all drilldowns into an array and do nothing if there are none
      var drilldowns = [];
      $.each($('input[name=drilldown]'), function(idx, item) {
        drilldowns.push($(this).val());
      });
      if (drilldowns.length === 0)
        return false;

      var year = $('#year').val();

      var shortcode_attrs = [' type="', visualisation, '"',
                             ' dataset="', dataset, '"',
                             ' drilldowns="', drilldowns.join(','), '"',
                            ];
      if (year !== '') {
        shortcode_attrs.push(' year="'+year+'"');
      }

      // Since we have everything we add the shortcode to page/post
      tinyMCEPopup.editor.execCommand('mceInsertContent', false,
        '[openspending'+shortcode_attrs.join('')+']');

      // Close the popup
      tinyMCEPopup.close();
      return false;
    });

    // Fetch buttons play a big part so we'll just search for them once
    var $drilldown_fetcher = $('#fetch-drilldowns');
    var $year_fetcher = $('#fetch-years');

    // When a user presses a key in dataset we activate the fetch drilldown
    // button since the user is now able to fetch drilldowns
    $('#dataset').keypress(function(e) {
        $drilldown_fetcher.removeClass('button-disabled');
        $year_fetcher.removeClass('button-disabled');
    });

    // When a user clicks the button to fetch drilldowns a lot of stuff happens
    $drilldown_fetcher.click(function(e) {
      // We add a spinner/loading gif to dimensions (to show we're working)
      var $dimensions = $('#dimensions');
      $dimensions.html('<img src="<?php print_asset('spinner'); ?>">');

      // Get the dataset, if the dataset field is empty (for example if the user
      // pressed the deactivated button) we tell the user, no dataset was
      // supplied
      var dataset = $('#dataset').val();
      if (dataset === '') {
          $dimensions.text('No dataset supplied');
      }
      // But if it has been supplied, we go out and fetch it's dimensions
      else {
        // Get the dimension via a specific call to openspending.org
        $.getJSON('https://openspending.org/'+dataset+'/dimensions.json')
          .done(function(data) { // Successful fetching
             // Disable the fetching button (it's been used)
            $drilldown_fetcher.addClass('button-disabled');
              // Create an array of possible dimensions and fill it with
              // Dimension links (a tags). The links store the key itself
              // in a data attribute
              var possible_dimensions = [];
              $.each(data, function(idx, dimension) {
                var dimension_link = ['<a href="#" data-dimension="',
                                      dimension.key,
                                      '" class="possible-dimension">',
                                      dimension.label,
                                      '</a>'
                                     ];
                possible_dimensions.push(dimension_link.join(''));
              });

              // Sort the dimensions for the user
              possible_dimensions.sort(function(a,b) {
                // We sort them according to labels, not dimension.keys
                var label_a = a.split('>')[1];
                var label_b = b.split('>')[1];

                if (label_a < label_b) return -1;
                if (label_a > label_b) return 1;
                return 0;
              });

              // Add the possible dimensions along with a header to the page
              $dimensions.append('Choose dimensions in drilldown order<br/>');
              $dimensions.html(possible_dimensions.join(', '));
            })
            .fail(function() { // Something went boo boo!
              // Let the user know
              $dimensions.text('Could not fetch dimensions of dataset');
            });

        // Possible dimensions are clickable (but since they're added live
        // we need to treat them as such)
        $('.possible-dimension').live('click', function(e) {
          // We add the clicked dimension to a chosen dimensions field
          var $dimension = $(this);
          var $chosen = $('#chosen-dimensions');
          $('#dimension-label').show();

          // The real input of the form (we could use something simpler but
          // since we need something to hold the value, why not the proper
          // one)?
          var $input = $(['<input name="drilldown" type="hidden" value="',
                          $dimension.attr('data-dimension'),
                          '">'].join(''));

          // The label to show the user
          var $text = $('<span class="dimension-label"/>')
                        .append($dimension.text());

          // Make it possible for the user to remove the chosen dimension
          $erase = $('<a href="#" class="delete">&times;</a>');

          // Add the choice to the chosen dimensions
          $choice = $('<span class="chosen-dimension"/>')
                      .append($text)
                      .append($input)
                      .append($erase);

          $chosen.append($choice);

          // Since we've added something now we can activate the
          // "Add visualisations button"
          $('#add-visualisation').removeClass('button-primary-disabled');
        });

        // User can remove a chosen dimension (which is added live)
        $('.delete').live('click', function(e) {
          $(this).parent().remove();
          return false;
        });
      }
      return false;
    });

    // When a user clicks the button to fetch years we populate the field
    $year_fetcher.click(function(e) {
      // We add a spinner/loading gif to years (to show we're working)
      var $years = $('#years');
      $years.html('<img src="<?php print_asset('spinner'); ?>">');

      // Get the dataset, if the dataset field is empty (for example if the user
      // pressed the deactivated button) we tell the user, no dataset was
      // supplied
      var dataset = $('#dataset').val();
      if (dataset === '') {
        $years.text('No dataset supplied');
      }
      else {
        // Get the dimension via a specific call to openspending.org
        $.getJSON('https://openspending.org/'+dataset+'/time.distinct')
          .done(function(data) { // Successful fetching
            // Disable the fetching button (it's been used)
            $year_fetcher.addClass('button-disabled');
            // Create an array of possible years and fill it with
            // Year links (a tags).
            var possible_years = [];
            // Since the data can have many dates per year and we're only
            // interested in the year, we create a "set" (object) to hold
            // the years (years are found in year element of each obj
            // in the data.results)
            var year_data = {};
            $.each(data.results, function(idx, time) {
              year_data[time.year] = true;
            });

            // Now we can loop through the keys of year_data to get the
            // unique years
            $.each(year_data, function(year, value) {
              var year_link = ['<a href="#" class="possible-year">',
                               year, '</a>'];
              possible_years.push(year_link.join(''));
            });

            // Sort the years for better user experience
            possible_years.sort()

            // Add the possible dimensions along with a header to the page
            $years.append('Click on the year to use<br/>');
            $years.html(possible_years.join(' '));
          })
          .fail(function() { // Something went boo boo!
            // Let the user know
            $years.text('Could not fetch years for the dataset');
          });
      }
      $('.possible-year').live('click', function(e) {
        var $year = $(this);
        // Unselect if previously chosen
        if ($year.hasClass('chosen')) {
          // Not chosen anymore
          $year.removeClass('chosen');
          // Reset the value
          $('input[name="year"]').val('')
        }
        // Select this year
        else {
          // Unselect other years
          $('.possible-year').removeClass('chosen');
          // Select this one and use its value
          $year.addClass('chosen');
          $('input[name="year"]').val($year.text());
        }

        return false;
      });

      return false;
    });
  })(jQuery);
</script>
</body>
</html>
